src/client.ts

```diff
@@ -43,39 +32,93 @@ export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
 export class APIClient {
   constructor(readonly config: IntegrationConfig) {}
 
-  public async verifyAuthentication(): Promise<void> {
-    // TODO make the most light-weight request possible to validate
-    // authentication works with the provided credentials, throw an err if
-    // authentication fails
-    const request = new Promise<void>((resolve, reject) => {
-      http.get(
-        {
-          hostname: 'localhost',
-          port: 443,
-          path: '/api/v1/some/endpoint?limit=1',
-          agent: false,
-          timeout: 10,
-        },
-        (res) => {
-          if (res.statusCode !== 200) {
-            reject(new Error('Provider authentication failed'));
-          } else {
-            resolve();
-          }
-        },
-      );
+  private withBaseUri(path: string): string {
+    return `https://${this.config.cybereasonHost}:${this.config.cybereasonPort}/${path}`;
+  }
+
+  private async getRequest(uri: string, cookie: string): Promise<Response> {
+    const response = await fetch(uri, {
+      headers: {
+        cookie,
+      },
     });
+    if (!response.ok) {
+      throw new IntegrationProviderAPIError({
+        endpoint: uri,
+        status: response.status,
+        statusText: response.statusText,
+      });
+    }
+    return response;
+  }
 
-    try {
-      await request;
-    } catch (err) {
-      throw new IntegrationProviderAuthenticationError({
-        cause: err,
-        endpoint: 'https://localhost/api/v1/some/endpoint?limit=1',
-        status: err.status,
-        statusText: err.statusText,
+  private async postRequest(
+    uri: string,
+    body: any,
+    dontRedirect?: boolean,
+  ): Promise<Response> {
+    const redirect = dontRedirect ? 'manual' : 'follow';
+
+    const response = await fetch(uri, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: querystring.stringify(body),
+      redirect,
+    });
+    if (!(response.ok || (response.status === 302 && dontRedirect))) {
+      throw new IntegrationProviderAPIError({
+        endpoint: uri,
+        status: response.status,
+        statusText: response.statusText,
       });
     }
+    return response;
+  }
+
+  public async login(): Promise<string> {
+    const loginRoute = this.withBaseUri('login.html');
+    const res = await this.postRequest(
+      loginRoute,
+      {
+        username: this.config.cybereasonId,
+        password: this.config.cybereasonPassword,
+      },
+      true,
+    );
+    return (
+      res.headers
+        .get('set-cookie')
+        ?.split('; ')
+        .filter((item) => item.search('JSESSIONID') >= 0)[0] || ''
+    );
+  }
+
+  public async logout(cookie: string): Promise<void> {
+    const logoutRoute = this.withBaseUri('logout');
+    await this.getRequest(logoutRoute, cookie);
+  }
+
+  public async verifyAuthentication(): Promise<void> {
```

Comment:
Auth verification flow:

login() -> test cookie against an api (proceed if successful, else throw Auth error) -> logout()

Reply:
Note: This would eventually be done on every API request since as discussed, we can't use setData. Will probably move this to a reusable method down the line.

Reply:
This is a good direction! Here's a suggestion:

1. Create something like `private static authCookie: string` as a Class' member.
2. Then your `login()` method will have a function of getting the auth cookie, a better name might be `getAuthCookie()`
3. In your request methods (`getRequest()`, `postRequest()` *), you can check if the cookie is already fetched before your `fetch()` call:
```
if (!APIClient.authCookie) {
  await this.getAuthCookie();
}
```

Since every call will go through the request method, its responsibility will be to check if the cookie exists and if not to fetch it. What do you think about this? @RPGPH 

Here's a good example of this in action:
1) https://github.com/JupiterOne/graph-checkmarx/blob/master/src/client.ts#L33 (Also please have a similar comment)
2) https://github.com/JupiterOne/graph-checkmarx/blob/master/src/client.ts#L65
3) https://github.com/JupiterOne/graph-checkmarx/blob/master/src/client.ts#L50

*) Also, can we have a single `request` method that can do both POST and GET requests instead of two, as seen here: https://github.com/JupiterOne/graph-checkmarx/blob/master/src/client.ts#L47?